File: solutions/__112_is_balance_tree.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func abs(a, b int) int {
	if a-b < 0 {
		return b - a
	}
	return a - b
}

func getDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	return max(getDepth(root.Left), getDepth(root.Right)) + 1
}

func isBalanced(root *TreeNode) bool {
	if root == nil {
		return true
	}
	lh := getDepth(root.Left)
	rh := getDepth(root.Right)
	if abs(lh, rh) > 1 {
		return false
	}
	return isBalanced(root.Left) && isBanlanced(root.Right)
}

File: solutions/__112_is_balance_tree.java
public class __112_is_balance_tree {
    public static void main(String[] args) {

    }

    public boolean isBalanced(TreeNode root) {
        if (root == null) {
            return true;
        }
        int left_depth = depth(root.left);
        int right_depth = depth(root.right);
        if (Math.abs(left_depth-right_depth) > 1) {
            return false;
        }
        return isBalanced(root.left) && isBalanced(root.right);
    }

    private int depth(TreeNode root) {
        if (root == null) {
            return -1;
        }
        return Math.max(depth(root.left), depth(root.right)) + 1;
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
